### PR TITLE
Removed old main cluster and renamed new main cluster in replicated upgrade

### PR DESCRIPTION
**Pull request by cchen-vertica**
This PR added the last two steps of replicated upgrade: 
1. remove the old main cluster. After we promote the sandbox, the old main cluster is not needed anymore so we remove it.
2. rename the subclusters in new main cluster. After we promote the sandbox, we rename all subclusters in the sandbox to match the ones in original main cluster.

### CODE REVIEW

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go [hunk 1]
```diff
@@ -929,3 +968,176 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
...
+	r.Manager.traceActorReconcile(actor)
+	return actor.Reconcile(ctx, &ctrl.Request{})
+}
+
+// renameReplicaGroupBFromVdb will rename the subclusters in promoted-sandbox/new-main-cluster to
+// match the ones in original main cluster
+func (r *ReplicatedUpgradeReconciler) renameReplicaGroupBFromVdb(ctx context.Context) (ctrl.Result, error) {
+	// if replica group A still exists or subclusters in replica group B have been renamed, we skip
+	// renaming the subclusters in replica group B
+	if r.countSubclustersForReplicaGroup(vmeta.ReplicaGroupAValue) > 0 ||
```
Comment (roypaulin) [inline pkg/controllers/vdb/replicatedupgrade_reconciler.go:1067]: If replicaA still exists, we should requeue.

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go [hunk 2]
```diff
@@ -929,3 +968,176 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
...
+	if !found {
+		r.Log.Info("Requeue because there are no primary UP nodes in main cluster to execute rename-subcluster operation")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	scNames := r.getSubclustersForReplicaGroup(vmeta.ReplicaGroupBValue)
+	for _, scName := range scNames {
+		newScName, found := r.scNameMap[scName]
+		// ignore the subcluster that has been renamed
+		if !found {
```
Comment (roypaulin) [inline pkg/controllers/vdb/replicatedupgrade_reconciler.go:1082]: Does not found mean it has already been renamed?

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go [hunk 3]
```diff
@@ -929,3 +968,176 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
...
+	return nil
+}
+
+// updateSubclusterNamesInVdb will update the names of subclusters in VerticaDB
+func (r *ReplicatedUpgradeReconciler) updateSubclusterNamesInVdb(ctx context.Context, scName, newScName string) error {
+	r.Log.Info("starting renaming subcluster in VerticaDB", "subcluster", scName, "new subcluster name", newScName)
+
+	updateSubclustersInVdb := func() (bool, error) {
+		// rename subcluster
+		for i := range r.VDB.Spec.Subclusters {
```
Comment (roypaulin) [inline pkg/controllers/vdb/replicatedupgrade_reconciler.go:1126]: use `vdb.GenSubclusterMap`.

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go [hunk 4]
```diff
@@ -929,3 +968,176 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
...
+			if r.VDB.Spec.Subclusters[i].Name == scName {
+				r.VDB.Spec.Subclusters[i].Name = newScName
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+
+	updated, err := vk8s.UpdateVDBWithRetry(ctx, r.VRec, r.VDB, updateSubclustersInVdb)
```
Comment (roypaulin) [inline pkg/controllers/vdb/replicatedupgrade_reconciler.go:1135]: Not related to this but in sandbox_reconciler vdb is updated before the sandbox api call. Can you change that?

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go (hunk 2 above)
Comment (cchen-vertica) [reply pkg/controllers/vdb/replicatedupgrade_reconciler.go:1082]: Yes, if the subcluster in group B cannot be found in scNameMap, it means that subcluster is renamed or removed. However, we won't remove subclusters in group B in replicated upgrade. So, we know the subcluster is renamed at this point.

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go (hunk 3 above)
Comment (cchen-vertica) [reply pkg/controllers/vdb/replicatedupgrade_reconciler.go:1126]: GenSubclusterMap needs to do a full iteration on all subclusters which is slow. Here if I found the right subcluster, I can exit.

File: pkg/controllers/vdb/replicatedupgrade_reconciler.go [hunk 5]
```diff
@@ -940,3 +977,164 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
...
+		}
+		return removed, nil
+	}
+
+	updated, err := vk8s.UpdateVDBWithRetry(ctx, r.VRec, r.VDB, updateSubclustersInVdb)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to delete subclusters in old main cluster: %w", err)
+	}
+	if updated {
+		r.Log.Info("deleted subclusters in old main cluster", "subclusters", scNames)
```
Comment (roypaulin) [inline pkg/controllers/vdb/replicatedupgrade_reconciler.go:1018]: I have not deleted the old main yet. The nodes are still up and scs still exist.

**Review by roypaulin — COMMENTED** [review]
We also need to clean up some annotations like `vertica.com/child-subcluster` when upgrade has finished.

File: pkg/controllers/vdb/promotesandboxtomain_reconciler.go [hunk 6]
```diff
@@ -174,6 +175,8 @@ func (s *PromoteSandboxToMainReconciler) updateSandboxInVdbStatus(ctx context.Co
 			vdbChg.Status.Sandboxes = append(vdbChg.Status.Sandboxes[:i], vdbChg.Status.Sandboxes[i+1:]...)
 			break
 		}
+		// update the annotation to show promote_sandbox has been done
+		vdbChg.Annotations[vmeta.ReplicatedUpgradeSandboxPromotedAnnotation] = vmeta.SandboxPromotedTrue
```
Comment (roypaulin) [inline pkg/controllers/vdb/promotesandboxtomain_reconciler.go:179]: I think this should be in `replicatedupgrade_reconciler` in the promotesandbox function, that's the only place we need it. And this function is to update the status  so not a good place to update the annotation.

File: pkg/controllers/vdb/dbremovesubcluster_reconciler.go [hunk 7]
```diff
@@ -180,6 +203,24 @@ func (d *DBRemoveSubclusterReconciler) updateSubclusterStatus(ctx context.Contex
 	return vdbstatus.Update(ctx, d.VRec.Client, d.Vdb, refreshInPlace)
 }
 
+// updateAnnotationForReplicatedUpgrade updates the annotation ReplicatedUpgradeReplicaARemovedAnnotation
+// to indicate we have removed the subclusters in replica A successfully
+func (d *DBRemoveSubclusterReconciler) updateAnnotationForReplicatedUpgrade(ctx context.Context) error {
```
Comment (roypaulin) [inline pkg/controllers/vdb/dbremovesubcluster_reconciler.go:208]: This should be in `replicatedupgrade_reconciler just after `removesubcluster`.
Also, I suggest you use `vk8s.MetaUpdate` to update annotations.

File: pkg/controllers/vdb/dbremovesubcluster_reconciler.go [hunk 8]
```diff
@@ -87,7 +91,15 @@ func (d *DBRemoveSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Re
 		return ctrl.Result{Requeue: changed}, err
 	}
 
-	return d.removeExtraSubclusters(ctx)
+	if res, err := d.removeExtraSubclusters(ctx); verrors.IsReconcileAborted(res, err) {
+		return res, err
+	}
+
+	if d.CalledInReplicatedUpgrade {
```
Comment (roypaulin) [inline pkg/controllers/vdb/dbremovesubcluster_reconciler.go:98]: If you move annotations' update to replicatedupgrade you won't need this

**Comment by roypaulin** [comment]
You should also clean up subcluster annotations that were added during upgrade.

File: pkg/controllers/vdb/dbremovesubcluster_reconciler.go (hunk 7 above)
Comment (cchen-vertica) [reply pkg/controllers/vdb/dbremovesubcluster_reconciler.go:208]: Why `vk8s.MetaUpdate` is better? Does it have a retry logic?

**Comment by cchen-vertica** [comment]
> You should also clean up subcluster annotations that were added during upgrade.

I've done that in upgrade.go. Did I miss some subcluster annotations?

**Comment by roypaulin** [comment]
> > You should also clean up subcluster annotations that were added during upgrade.
> 
> I've done that in upgrade.go. Did I miss some subcluster annotations?

What you have updated are vdb annotations. I am talking about subcluster annotations. Some of the ones that are added in `duplicateSubclusterForReplicaGroupB`

**Comment by cchen-vertica** [comment]
> > > You should also clean up subcluster annotations that were added during upgrade.
> > 
> > 
> > I've done that in upgrade.go. Did I miss some subcluster annotations?
> 
> What you have updated are vdb annotations. I am talking about subcluster annotations. Some of the ones that are added in `duplicateSubclusterForReplicaGroupB`

They have been removed in line 267,268 of upgrade.go. Are 'ParentSubclusterTypeAnnotation' and 'ParentSubclusterAnnotation' you are referring to?

**Review by roypaulin — APPROVED** [review]

**Pull request merged**